static/app/views/organizationContext.tsx

```diff
@@ -1,5 +1,150 @@
+  };
+
+  getChildContext() {
+    return {organization: this.props.value};
+  }
+
+  render() {
+    return this.props.children;
+  }
+}
+
+/**
+ * Context provider responsible for loading the organization into the
+ * OrganizationStore if it is not already present.
+ */
+export function OrganizationContextProvider({children}: Props) {
+  const api = useApi();
+  const configStore = useLegacyStore(ConfigStore);
+
+  const {organizations} = useLegacyStore(OrganizationsStore);
+  const {organization, error} = useLegacyStore(OrganizationStore);
+
+  const hasMadeFirstFetch = useRef(false);
+
+  const lastOrganizationSlug: string | null =
+    configStore.lastOrganization ?? organizations[0]?.slug ?? null;
+
+  const routes = useRoutes();
+  const params = useParams<{orgId?: string}>();
+
+  // XXX(epurkhiser): When running in deploy preview mode customer domains are
+  // not supported correctly. Do NOT use the customer domain from the params.
+  const orgSlug = DEPLOY_PREVIEW_CONFIG
+    ? lastOrganizationSlug
+    : params.orgId || lastOrganizationSlug;
+
+  // If the organization slug differs from what we have in the organization
+  // store reload the store
+  useEffect(() => {
+    // Nothing to do if we already have the organization loaded
+    if (organization && organization.slug === orgSlug) {
+      return;
+    }
+
+    if (!orgSlug) {
+      return;
+    }
+
+    metric.mark({name: 'organization-details-fetch-start'});
+    fetchOrganizationDetails(api, orgSlug, false, hasMadeFirstFetch.current);
+    hasMadeFirstFetch.current = true;
+  }, [api, orgSlug, organization]);
+
+  // Take a measurement for when organization details are done loading and the
+  // new state is applied
+  useEffect(
+    () => {
+      if (organization === null) {
+        return;
+      }
+
+      metric.measure({
+        name: 'app.component.perf',
+        start: 'organization-details-fetch-start',
+        data: {
+          name: 'org-details',
+          route: getRouteStringFromRoutes(routes),
+          organization_id: parseInt(organization.id, 10),
+        },
+      });
+    },
+    // Ignore the `routes` dependency for the metrics measurement
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [organization]
+  );
+
+  const {user} = configStore;
+
+  // If we've had an error it may be possible for the user to use the sudo
+  // modal to load the organization.
+  useEffect(() => {
+    if (!error) {
+      return;
+    }
+
+    if (user.isSuperuser && error.status === 403) {
+      openSudo({isSuperuser: true, needsReload: true});
+    }
+
+    // This `catch` can swallow up errors in development (and tests)
+    // So let's log them. This may create some noise, especially the test case where
+    // we specifically test this branch
+    console.error(error); // eslint-disable-line no-console
+  }, [user, error]);
+
+  // Switch organizations when the orgId changes
+  const lastOrgId = useRef(orgSlug);
```

Comment:
A bit confusing how we have `lastOrganizationSlug` above and `lastOrgId` here? Also wondering if we can use the `usePrevious` hook here